### PR TITLE
feat: support mapping of legacy GCF events

### PR DESCRIPTION
**Pull request by coryan**
With this change the framework can receive GCF events and deliver them as Cloud Events.

Part of the work for #306

### CODE REVIEW

**Review by devjgm — APPROVED** [review]
LGTM, just a couple small questions.

File: google/cloud/functions/cloud_event.cc [hunk 1]
```diff
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 #include "google/cloud/functions/cloud_event.h"
-#include <absl/time/time.h>
+#include <absl/time/time.h>  // NOLINT(modernize-deprecated-headers)
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/functions/cloud_event.cc:16]: I agree.  Whatever regular expression they are using, it is a bit off.

File: google/cloud/functions/internal/parse_cloud_event_legacy.cc [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+  }();
+  auto gcf_event_id = GetAlternatives(json, {"context", "eventId"}, "eventId");
+  auto gcf_resource_name =
+      GetAlternatives(json, {"context", "resource", "name"}, "resource");
+  auto gcf_timestamp =
+      GetAlternatives(json, {"context", "timestamp"}, "timestamp");
+  auto gcf_source = "//" + gcf_service + "/" + gcf_resource_name;
+
+  return LegacyCommonFields{/*.event_type=*/std::move(gcf_event_type),
+                            /*.service=*/std::move(gcf_service),
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/functions/internal/parse_cloud_event_legacy.cc:137]: Ack.

File: google/cloud/functions/internal/parse_cloud_event_legacy.cc [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+                            std::deque<std::string> primary,
+                            std::string const& alternative) {
+  auto value = GetNestedKey(json, std::move(primary));
+  if (!value.empty()) return value;
+  return json.value(alternative, "");
+}
+
+std::string MapGCFTypeToService(std::string const& gcf_event_type) {
+  static auto const* const kPrefixes =
+      new auto(std::vector<std::pair<std::string, std::string>>{
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/functions/internal/parse_cloud_event_legacy.cc]: Fixed.

File: google/cloud/functions/internal/parse_cloud_event_legacy.cc [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+  return event;
+}
+
+functions::CloudEvent ParseLegacyStorage(nlohmann::json const& json,
+                                         LegacyCommonFields gcf) {
+  auto const re = std::regex(
+      "//storage\\.googleapis\\.com/"
+      "projects/_/buckets/([^/]+)/objects/(.+)");
+  std::smatch m;
+  if (std::regex_match(gcf.source, m, re) && m.size() > 2) {
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/functions/internal/parse_cloud_event_legacy.cc]: Fixed.

**Pull request merged**